File: server/wiresharkServer.ts
import express from 'express';
import { createServer } from 'http';
import { Server } from 'socket.io';
import { spawn } from 'child_process';
import cors from 'cors';
import { WiresharkData, NetworkHost, NetworkStream } from '../types/wireshark';

const app = express();
app.use(cors());

const httpServer = createServer(app);
const io = new Server(httpServer, {
  cors: {
    origin: ["http://localhost:5173", "http://localhost:5174", "http://localhost:5175"],
    methods: ["GET", "POST"],
    credentials: true
  },
  pingTimeout: 60000,
  pingInterval: 25000,
  transports: ['websocket', 'polling']
});

// Add health check endpoint
app.get('/health', (req, res) => {
  res.json({ status: 'ok' });
});

interface PacketData {
  _source: {
    layers: {
      frame: { "frame.time_epoch": string };
      ip: { 
        "ip.src": string;
        "ip.dst": string;
        "ip.proto": string;
        "ip.len": string;
      };
      tcp?: { "tcp.srcport": string; "tcp.dstport": string };
      udp?: { "udp.srcport": string; "udp.dstport": string };
    };
  };
}

class NetworkTrafficAggregator {
  private hosts: Map<string, NetworkHost> = new Map();
  private streams: Map<string, NetworkStream> = new Map();
  private hostIdCounter = 0;

  addPacket(packet: PacketData) {
    const { ip } = packet._source.layers;
    const srcIp = ip["ip.src"];
    const dstIp = ip["ip.dst"];
    const protocol = this.getProtocol(packet);
    const bytes = parseInt(ip["ip.len"]);
    const timestamp = parseFloat(packet._source.layers.frame["frame.time_epoch"]) * 1000;

    // Update hosts
    const srcHost = this.getOrCreateHost(srcIp);
    const dstHost = this.getOrCreateHost(dstIp);

    srcHost.packets++;
    dstHost.packets++;
    srcHost.bytesTransferred += bytes;
    dstHost.bytesTransferred += bytes;

    // Update stream
    const streamKey = `${srcHost.id}-${dstHost.id}-${protocol}`;
    let stream = this.streams.get(streamKey);
    if (!stream) {
      stream = {
        source: srcHost.id,
        target: dstHost.id,
        protocol,
        packets: 0,
        bytes: 0,
        timestamp
      };
      this.streams.set(streamKey, stream);
    }
    stream.packets++;
    stream.bytes += bytes;
    stream.timestamp = timestamp;

    return this.getVisualizationData();
  }

  private getOrCreateHost(ip: string): NetworkHost {
    let host = Array.from(this.hosts.values()).find(h => h.ip === ip);
    if (!host) {
      host = {
        id: (++this.hostIdCounter).toString(),
        ip,
        packets: 0,
        bytesTransferred: 0
      };
      this.hosts.set(host.id, host);
    }
    return host;
  }

  private getProtocol(packet: PacketData): string {
    const { ip } = packet._source.layers;
    if (packet._source.layers.tcp) return 'TCP';
    if (packet._source.layers.udp) return 'UDP';
    return 'OTHER';
  }

  getVisualizationData(): WiresharkData {
    return {
      hosts: Array.from(this.hosts.values()),
      streams: Array.from(this.streams.values())
    };
  }
}

const startCapture = (networkInterface: string = 'any') => {
  const aggregator = new NetworkTrafficAggregator();
  
  // Use sudo to run tshark with elevated privileges
  const tshark = spawn('sudo', [
    'tshark',
    '-i', networkInterface,
    '-T', 'json',
    '-l',
    '-f', 'ip'  // Filter for IP packets only
  ]);

  tshark.stderr.on('data', (data) => {
    const errorMsg = data.toString();
    console.error('tshark error:', errorMsg);
    if (errorMsg.includes('Permission denied')) {
      io.emit('error', { 
        message: 'Permission denied. Please run the server with sudo privileges.'
      });
    }
  });

  let buffer = '';
  tshark.stdout.on('data', (data) => {
    buffer += data.toString();
    
    try {
      const packet: PacketData = JSON.parse(buffer);
      const visualizationData = aggregator.addPacket(packet);
      io.emit('networkUpdate', visualizationData);
      buffer = '';
    } catch (e) {
      // Incomplete JSON, continue buffering
      if (buffer.length > 1000000) { // Reset if buffer gets too large
        buffer = '';
      }
    }
  });

  tshark.on('error', (error) => {
    console.error('Failed to start tshark:', error);
    io.emit('error', { 
      message: 'Failed to start packet capture. Check permissions and tshark installation.'
    });
  });

  return tshark;
};

// Mock data generator for test traffic
class TestTrafficGenerator {
  private running = false;
  private interval: NodeJS.Timeout | null = null;
  private aggregator: NetworkTrafficAggregator;
  private ips = [
    '192.168.1.1', '192.168.1.2', '192.168.1.100', 
    '10.0.0.1', '10.0.0.2', '10.0.0.3',
    '172.16.0.1', '8.8.8.8', '1.1.1.1'
  ];
  private protocols = ['TCP', 'UDP', 'HTTP', 'HTTPS', 'DNS'];
  private connectionAttempts = 0;
  private maxRetries = 5;
  
  constructor() {
    this.aggregator = new NetworkTrafficAggregator();
  }

  start(io: Server) {
    if (this.running) return;
    this.running = true;
    
    // Reset connection attempts
    this.connectionAttempts = 0;
    
    console.log('Starting test traffic generator with enhanced reliability');
    
    // Generate traffic every 500ms
    this.interval = setInterval(() => {
      try {
        if (io.engine && io.engine.clientsCount > 0) {
          const packet = this.generateRandomPacket();
          const visualizationData = this.aggregator.addPacket(packet);
          io.emit('networkUpdate', visualizationData);
        } else {
          console.log('No clients connected, waiting for connections...');
          this.connectionAttempts++;
          
          if (this.connectionAttempts > this.maxRetries) {
            console.log(`Exceeded max retry attempts (${this.maxRetries}). Pausing traffic generation.`);
            this.pause();
          }
        }
      } catch (error) {
        console.error('Error generating test traffic:', error);
        // Don't stop completely on error, just log it
      }
    }, 500);
  }

  pause() {
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = null;
    }
  }

  resume(io: Server) {
    if (this.running && !this.interval) {
      this.connectionAttempts = 0;
      this.start(io);
    }
  }

  stop() {
    if (!this.running) return;
    this.running = false;
    
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = null;
    }
    
    // Reset the aggregator to clear data
    this.aggregator = new NetworkTrafficAggregator();
  }

  private generateRandomPacket(): PacketData {
    const srcIpIndex = Math.floor(Math.random() * this.ips.length);
    let dstIpIndex = Math.floor(Math.random() * this.ips.length);
    
    // Ensure source and destination are different
    while (dstIpIndex === srcIpIndex) {
      dstIpIndex = Math.floor(Math.random() * this.ips.length);
    }
    
    const protocolIndex = Math.floor(Math.random() * this.protocols.length);
    const protocol = this.protocols[protocolIndex];
    const bytes = Math.floor(Math.random() * 1500) + 50; // Random packet size between 50 and 1550 bytes
    
    let tcpData;
    let udpData;
    
    if (protocol === 'TCP' || protocol === 'HTTP' || protocol === 'HTTPS') {
      const srcPort = (Math.floor(Math.random() * 60000) + 1024).toString();
      let dstPort = (Math.floor(Math.random() * 60000) + 1024).toString();
      
      // Set appropriate destination ports for HTTP/HTTPS
      if (protocol === 'HTTP') {
        dstPort = '80';
      } else if (protocol === 'HTTPS') {
        dstPort = '443';
      }
      
      tcpData = { 
        "tcp.srcport": srcPort,
        "tcp.dstport": dstPort
      };
    } else if (protocol === 'UDP' || protocol === 'DNS') {
      const srcPort = (Math.floor(Math.random() * 60000) + 1024).toString();
      let dstPort = (Math.floor(Math.random() * 60000) + 1024).toString();
      
      // Set appropriate destination port for DNS
      if (protocol === 'DNS') {
        dstPort = '53';
      }
      
      udpData = {
        "udp.srcport": srcPort,
        "udp.dstport": dstPort
      };
    }
    
    return {
      _source: {
        layers: {
          frame: { 
            "frame.time_epoch": (Date.now() / 1000).toString()
          },
          ip: {
            "ip.src": this.ips[srcIpIndex],
            "ip.dst": this.ips[dstIpIndex],
            "ip.proto": protocol === 'TCP' || protocol === 'HTTP' || protocol === 'HTTPS' ? '6' : 
                        protocol === 'UDP' || protocol === 'DNS' ? '17' : '1',
            "ip.len": bytes.toString()
          },
          tcp: tcpData,
          udp: udpData
        }
      }
    };
  }
}

// Create test traffic generator
const testTrafficGenerator = new TestTrafficGenerator();

io.on('connection', (socket) => {
  console.log('Client connected');
  let capture: ReturnType<typeof startCapture> | null = null;
  
  socket.on('startCapture', (networkInterface: string) => {
    console.log('Starting capture on interface:', networkInterface);
    try {
      if (networkInterface === 'test') {
        console.log('Starting test traffic generator');
        testTrafficGenerator.start(io);
      } else {
        capture = startCapture(networkInterface);
      }
    } catch (err) {
      console.error('Failed to start capture:', err);
      socket.emit('error', { message: 'Failed to start capture' });
    }
  });

  socket.on('stopTestTraffic', () => {
    console.log('Stopping test traffic generator');
    testTrafficGenerator.stop();
  });

  socket.on('error', (error) => {
    console.error('Socket error:', error);
  });
  
  socket.on('disconnect', () => {
    console.log('Client disconnected');
    if (capture) {
      capture.kill();
      console.log('Capture stopped');
    }
    
    // Also stop test traffic generator if running
    testTrafficGenerator.stop();
  });
});

const cleanup = () => {
  console.log('Cleaning up...');
  // Close connections gracefully
  io.close(() => {
    httpServer.close(() => {
      console.log('Server shut down successfully');
      // Only exit if explicitly requested via SIGINT or SIGTERM
    });
  });
};

// Handle termination signals
process.on('SIGINT', () => {
  console.log('Received SIGINT signal');
  cleanup();
  process.exit(0);
});

process.on('SIGTERM', () => {
  console.log('Received SIGTERM signal');
  cleanup();
  process.exit(0);
});

// Add uncaught exception handler for better stability
process.on('uncaughtException', (error: Error) => {
  console.error('Uncaught exception:', error);
  
  // Handle port already in use errors specially
  if (error && 'code' in error && (error as NodeJS.ErrnoException).code === 'EADDRINUSE') {
    console.log(`Port ${PORT} is already in use. This could be another instance of the server.`);
    console.log('You can either:');
    console.log(`1. Use the existing server instance on port ${PORT}`);
    console.log('2. Stop the existing server and try again');
    console.log('3. Change the PORT environment variable to use a different port');
    
    // Exit with a special code for port in use
    process.exit(2);
  }
  
  // Don't exit immediately on other uncaught exceptions
  // This allows the server to keep running despite errors
});

const PORT = process.env.PORT || 3002;

// Try to start the server with graceful error handling
const startServer = () => {
  try {
    httpServer.listen(PORT, () => {
      console.log(`Server running on port ${PORT}`);
    });
  } catch (error) {
    console.error('Failed to start server:', error);
    // Error handling will be caught by uncaughtException handler
  }
};

startServer();

File: server/wiresharkServer.py
#!/usr/bin/env python3
import json
import signal
import subprocess
import sys
import time
from dataclasses import dataclass, field, asdict
from typing import Dict, List, Optional, Any, Tuple
import random

from flask import Flask, jsonify, request
from flask_cors import CORS
from flask_socketio import SocketIO, emit, disconnect

app = Flask(__name__)
CORS(app)
socketio = SocketIO(app, cors_allowed_origins=["http://localhost:5173", 
                                              "http://localhost:5174", 
                                              "http://localhost:5175"],
                   ping_timeout=60, ping_interval=25)

# Data models
@dataclass
class NetworkHost:
    id: str
    ip: str
    packets: int = 0
    bytesTransferred: int = 0

@dataclass
class NetworkStream:
    source: str
    target: str
    protocol: str
    packets: int = 0
    bytes: int = 0
    timestamp: float = 0.0

@dataclass
class WiresharkData:
    hosts: List[NetworkHost] = field(default_factory=list)
    streams: List[NetworkStream] = field(default_factory=list)

class NetworkTrafficAggregator:
    def __init__(self):
        self.hosts: Dict[str, NetworkHost] = {}
        self.streams: Dict[str, NetworkStream] = {}
        self.host_id_counter = 0

    def add_packet(self, packet: Dict[str, Any]) -> Dict[str, Any]:
        ip_layer = packet["_source"]["layers"]["ip"]
        src_ip = ip_layer["ip.src"]
        dst_ip = ip_layer["ip.dst"]
        protocol = self._get_protocol(packet)
        bytes_transferred = int(ip_layer["ip.len"])
        timestamp = float(packet["_source"]["layers"]["frame"]["frame.time_epoch"]) * 1000

        # Update hosts
        src_host = self._get_or_create_host(src_ip)
        dst_host = self._get_or_create_host(dst_ip)

        src_host.packets += 1
        dst_host.packets += 1
        src_host.bytesTransferred += bytes_transferred
        dst_host.bytesTransferred += bytes_transferred

        # Update stream
        stream_key = f"{src_host.id}-{dst_host.id}-{protocol}"
        if stream_key not in self.streams:
            self.streams[stream_key] = NetworkStream(
                source=src_host.id,
                target=dst_host.id,
                protocol=protocol,
                packets=0,
                bytes=0,
                timestamp=timestamp
            )
            
        stream = self.streams[stream_key]
        stream.packets += 1
        stream.bytes += bytes_transferred
        stream.timestamp = timestamp

        return self._get_visualization_data()

    def _get_or_create_host(self, ip: str) -> NetworkHost:
        for host in self.hosts.values():
            if host.ip == ip:
                return host
                
        self.host_id_counter += 1
        host = NetworkHost(
            id=str(self.host_id_counter),
            ip=ip
        )
        self.hosts[host.id] = host
        return host

    def _get_protocol(self, packet: Dict[str, Any]) -> str:
        layers = packet["_source"]["layers"]
        if "tcp" in layers:
            return "TCP"
        elif "udp" in layers:
            return "UDP"
        return "OTHER"

    def _get_visualization_data(self) -> Dict[str, Any]:
        return {
            "hosts": list(self.hosts.values()),
            "streams": list(self.streams.values())
        }


def start_capture(network_interface='any'):
    aggregator = NetworkTrafficAggregator()
    
    try:
        # Use tshark to capture network traffic
        tshark_process = subprocess.Popen(
            ["sudo", "tshark", "-i", network_interface, "-T", "json", "-l", "-f", "ip"],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )
        
        buffer = ""
        while True:
            output = tshark_process.stdout.readline()
            if output == "" and tshark_process.poll() is not None:
                break
                
            if output:
                buffer += output
                try:
                    packet = json.loads(buffer)
                    visualization_data = aggregator.add_packet(packet)
                    socketio.emit('networkUpdate', visualization_data)
                    buffer = ""
                except json.JSONDecodeError:
                    # Incomplete JSON, continue buffering
                    if len(buffer) > 1000000:  # Reset if buffer gets too large
                        buffer = ""
                        
            error = tshark_process.stderr.readline()
            if error:
                print(f"tshark error: {error}", file=sys.stderr)
                if "Permission denied" in error:
                    socketio.emit('error', {
                        "message": "Permission denied. Please run the server with sudo privileges."
                    })
                    
    except Exception as e:
        print(f"Failed to start tshark: {e}", file=sys.stderr)
        socketio.emit('error', {
            "message": "Failed to start packet capture. Check permissions and tshark installation."
        })
    finally:
        if tshark_process and tshark_process.poll() is None:
            tshark_process.terminate()
            tshark_process.wait()
            

class TestTrafficGenerator:
    def __init__(self):
        self.running = False
        self.aggregator = NetworkTrafficAggregator()
        self.ips = [
            '192.168.1.1', '192.168.1.2', '192.168.1.100', 
            '10.0.0.1', '10.0.0.2', '10.0.0.3',
            '172.16.0.1', '8.8.8.8', '1.1.1.1'
        ]
        self.protocols = ['TCP', 'UDP', 'HTTP', 'HTTPS', 'DNS']
        self.connection_attempts = 0
        self.max_retries = 5

    def generate_random_packet(self):
        src_ip_index = random.randint(0, len(self.ips) - 1)
        dst_ip_index = random.randint(0, len(self.ips) - 1)
        
        # Ensure source and destination are different
        while dst_ip_index == src_ip_index:
            dst_ip_index = random.randint(0, len(self.ips) - 1)
            
        protocol_index = random.randint(0, len(self.protocols) - 1)
        protocol = self.protocols[protocol_index]
        bytes_count = random.randint(50, 1550)  # Random packet size between 50 and 1550 bytes
        
        tcp_data = None
        udp_data = None
        
        if protocol in ['TCP', 'HTTP', 'HTTPS']:
            src_port = str(random.randint(1024, 61023))
            dst_port = str(random.randint(1024, 61023))
            
            # Set appropriate destination ports for HTTP/HTTPS
            if protocol == 'HTTP':
                dst_port = '80'
            elif protocol == 'HTTPS':
                dst_port = '443'
                
            tcp_data = {
                "tcp.srcport": src_port,
                "tcp.dstport": dst_port
            }
        elif protocol in ['UDP', 'DNS']:
            src_port = str(random.randint(1024, 61023))
            dst_port = str(random.randint(1024, 61023))
            
            # Set appropriate destination port for DNS
            if protocol == 'DNS':
                dst_port = '53'
                
            udp_data = {
                "udp.srcport": src_port,
                "udp.dstport": dst_port
            }
            
        return {
            "_source": {
                "layers": {
                    "frame": {
                        "frame.time_epoch": str(time.time())
                    },
                    "ip": {
                        "ip.src": self.ips[src_ip_index],
                        "ip.dst": self.ips[dst_ip_index],
                        "ip.proto": "6" if protocol in ['TCP', 'HTTP', 'HTTPS'] else
                                   "17" if protocol in ['UDP', 'DNS'] else "1",
                        "ip.len": str(bytes_count)
                    },
                    "tcp": tcp_data,
                    "udp": udp_data
                }
            }
        }

    def start(self):
        self.running = True
        self.connection_attempts = 0
        
        print("Starting test traffic generator with enhanced reliability")

# Health check endpoint
@app.route('/health')
def health():
    return jsonify({"status": "ok"})

# Socket.IO event handlers
@socketio.on('connect')
def handle_connect():
    print('Client connected')

@socketio.on('startCapture')
def handle_start_capture(network_interface):
    print(f'Starting capture on interface: {network_interface}')
    
    if network_interface == 'test':
        print('Starting test traffic generator')
        # Start test traffic generator in a background task
        socketio.start_background_task(start_test_traffic)
    else:
        # Start real traffic capture in a background task
        socketio.start_background_task(start_capture, network_interface)

@socketio.on('stopTestTraffic')
def handle_stop_test_traffic():
    print('Stopping test traffic generator')
    global test_traffic_generator
    test_traffic_generator.running = False

@socketio.on('disconnect')
def handle_disconnect():
    print('Client disconnected')
    
# Background task to generate test traffic
test_traffic_generator = TestTrafficGenerator()

def start_test_traffic():
    global test_traffic_generator
    test_traffic_generator.running = True
    test_traffic_generator.connection_attempts = 0
    
    while test_traffic_generator.running:
        try:
            packet = test_traffic_generator.generate_random_packet()
            visualization_data = test_traffic_generator.aggregator.add_packet(packet)
            socketio.emit('networkUpdate', visualization_data)
            socketio.sleep(0.5)  # Generate traffic every 500ms
        except Exception as e:
            print(f"Error generating test traffic: {e}")
            # Don't stop completely on error, just log it
            socketio.sleep(1)

# Signal handlers
def cleanup(signum, frame):
    print('Cleaning up...')
    print('Server shut down successfully')
    sys.exit(0)

signal.signal(signal.SIGINT, cleanup)
signal.signal(signal.SIGTERM, cleanup)

if __name__ == '__main__':
    PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 3002
    
    print(f"Server running on port {PORT}")
    socketio.run(app, host='0.0.0.0', port=PORT, debug=False)